hwinfo: cleanup API doc and test

alexanderwachter opened the pull request:
Change the doxygen doc from "negative on error" to -ENOSUP only.
Enhance testing.
Fix build-errors on NRF53, STM32H7, SAM3x, SAM4e, SAM4s

Fixes  #22979

alexanderwachter committed:
doc: hwinfo: Clarify return values.

Change the doxygen doc from "negative on error" to -ENOSUP and
negative value on driver specific errors.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

galak reviewed (changes requested):
We should probably fix the tests/drivers/hwinfo/api/src/main.c to reflect what's documented here.

ie, -ENOTSUP is a valid return and shouldn't assert.  An other negative value should assert since that would be a violation of the API.

alexanderwachter commented:
> We should probably fix the tests/drivers/hwinfo/api/src/main.c to reflect what's documented here.
> 
> ie, -ENOTSUP is a valid return and shouldn't assert. An other negative value should assert since that would be a violation of the API.

We only test on boards that have a hardware ID. It it returns -ENOSUP, we have a problem with the implementation.

galak commented:
> We only test on boards that have a hardware ID. It it returns -ENOSUP, we have a problem with the implementation.

If you want to handle that in the test I think its fine, you can add to Kconfig option to tell if a driver is enabled or not and thus distinguish if -ENOSUP is a valid response or not.


```
diff --git a/drivers/hwinfo/Kconfig b/drivers/hwinfo/Kconfig
index 173dfa9672..7b5f0c0b51 100644
--- a/drivers/hwinfo/Kconfig
+++ b/drivers/hwinfo/Kconfig
@@ -16,9 +16,13 @@ config HWINFO_SHELL
        help
          Enable hwinfo Shell for testing.
 
+config HWINFO_HAS_DRIVER
+       bool
+
 config HWINFO_STM32
        bool "STM32 hwinfo"
        default y
+       select HWINFO_HAS_DRIVER
        depends on SOC_FAMILY_STM32
        select USE_STM32_LL_UTILS
        help
@@ -27,6 +31,7 @@ config HWINFO_STM32
 config HWINFO_NRF
        bool "NRF device ID"
        default y
+       select HWINFO_HAS_DRIVER
        depends on SOC_FAMILY_NRF
        help
          Enable Nordic NRF hwinfo driver.
...
```

galak commented:
Than we can also get ride of 'hwinfo' from the board yaml and tests/drivers/hwinfo/api/testcase.yaml and change testcase.yaml filter on `CONFIG_HWINFO`

galak commented:
I'm fine approving and merging the doc update PR, just be aware I'd like to see the other changes in testing for us to have this considered a stable API.

alexanderwachter committed:
drivers: hwinfo: Introduce HWINFO_HAS_DRIVER Kconfig symbol

Introduce the HWINFO_HAS_DRIVER Kconfig symbol to get
the information if hwinfo is supported on the current platform.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
tests: hwinfo: Make the tests depending on HWINFO_HAS_DRIVER

The test checks for a correct implementation if HWINFO_HAS_DRIVER
is set, otherwise it checks for -ENOTSUP return value.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

galak commented on tests/drivers/hwinfo/api/testcase.yaml:
```diff
@@ -1,4 +1,3 @@
 tests:
   drivers.device_id:
     tags: driver
-    depends_on: hwinfo
```
make this `filter: CONFIG_HWINFO`

galak reviewed (changes requested):
looks good, one minor tweak on the test yaml

we should also remove `hwinfo` from all board.yaml that have it.

galak reviewed (commented):
On the 'Error returned:' in the test for negative retval, should we make that message a bit more verbose?

'Unexpected negative return value' or something?

alexanderwachter commented:
> we should also remove `hwinfo` from all board.yaml that have it.

Why should we remove it? It states which boards have drivers for hwinfo.

alexanderwachter committed:
drivers: hwinfo: Exclude NRF53 non-secure from hwinfo driver

Make the NRF hwinfo driver depending on !TRUSTED_EXECUTION_NONSECURE
because the FICR registers are not accessible from the non-secure
world.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
boards: Remove "supported: -hwinfo" from all boards

Remove all "supported: -hwinfo" definitions from the boards
yaml files and documentation. hwinfo can generally be tested
on every board because it returns -ENOTSUP if not supported.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter commented:
> > we should also remove `hwinfo` from all board.yaml that have it.
> 
> Why should we remove it? It states which boards have drivers for hwinfo.

@galak I think I understand now. The yamls are for tests only and with the new tests, we also test for not supported. That way we can test on every board.

alexanderwachter committed:
SoC: stm32: Include LL utilities if HWINFO is selected

Include stm32XXxx_ll_utils.h in soc.h for every stm32 SoC,
if CONFIG_HWINFO_STM32 is selected.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
drivers: hwinfo: Don't disable CLOE on SAM3x seies

CLOE (Code Loop Optimization) does not exist on SAM3x.
Make the EEFC_FMR_CLOE disable depending on CONFIG_SOC_SERIES_SAM3X.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter committed:
SoC: SAM3x and SAM4: Add the Flash Controller to the DT

This commit adds Device-Tree instances of the Flash controller
to the SAM3X, SAM4E and SAM4S series. The Flash-Controller
is used to get the unique device identifier.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

alexanderwachter commented:
Somebody needs to test the SAM3X and SAM4X hwinfo. I don't have any of these boards.

jfischer-phytec-iot commented on include/drivers/hwinfo.h:
```diff
@@ -39,7 +39,8 @@ extern "C" {
  *
- * @retval size of the device ID copied or negative on error.
+ * @retval size of the device ID copied
+ * @retval -ENOTSUP if there is no implementation for the particular device.
```
`size of the device ID copied or negative on error` is clear defined behavior, but not `-ENOTSUP if there is no implementation for the particular device`, what about other return values?

alexanderwachter replied:
There are no other errors.

jfischer-phytec-iot replied:
> There are no other errors.

What is the reason to change it? Is it sure that future implementations would not need to return anything else as -ENOTSUP? What about an internal error like -EIO or -EBUSY?

alexanderwachter replied:
I'd like to declare the API as stable, and the requirement from the API meeting was that I have to specify which errors could be returned. If we have another driver that could return another error, we have to announce this change, I guess.

alexanderwachter replied:
This PR started as doc-change and escalated quickly to 8 commits and +128 lines.

jfischer-phytec-iot replied:
I just do not see the reason to change clear defined `retval`.

alexanderwachter replied:
I was not in the meeting. I don‘t know what was discussed there. @cales might know.

galak replied:
From dev-review meeting: Document the specific error for -ENOTSUP and than document that any other negative value would be a driver specific error.

nandojve commented:
> Somebody needs to test the SAM3X and SAM4X hwinfo. I don't have any of these boards.
> @fallrisk @mnkp @nandojve

I help with tests on SAM4S/E variants. I don't have any SAM3 board.

ioannisg commented:
> > Somebody needs to test the SAM3X and SAM4X hwinfo. I don't have any of these boards.
> > @fallrisk @mnkp @nandojve
> 
> I help with tests on SAM4S/E variants. I don't have any SAM3 board.

I can test on SAM3X, hold on

alexanderwachter commented:
@nandojve @ioannisg can you guys check if the ID differs on two different boards? You can use the shell module to check. Just to make sure that I got the address right.

nandojve commented:
> @nandojve @ioannisg can you guys check if the ID differs on two different boards? You can use the shell module to check. Just to make sure that I got the address right.

Hi @alexanderwachter, I already tested on both boards. I separated values for better reading on the post. I added values that I get using Atmel Studio.
```
SAM4E
uart:~$ hwinfo devid 
Length: 16
ID: 0x00313353_50334337_30333032_30303436
[0] 0x53333100
[1] 0x37433350
[2] 0x32303330
[3] 0x36343030
uart:~$ 

SAM4S
uart:~$ hwinfo devid
Length: 16
ID: 0x00310053_594a5135_30323032_33303830
[0] 0x53003100
[1] 0x35514a59
[2] 0x32303230
[3] 0x30383033
uart:~$ 
```
Below I executed a demo inside Atmel Studio to get Device ID.
```
-- Flash Read Unique Identifier Example --
-- SAM4S-XPLAINED --
-- Compiled: Feb 20 2020 18:26:05 --
-I- Reading 128 bits Unique Identifier
	    [0]		[1]	  [2]	     [3]
-I- ID: 1392521472, 894519897, 842019376, 808988723
SAM4S
	0x53003100, 0x35514a59, 0x32303230, 0x30383033
SAM4E
        0x53333100, 0x37433350, 0x32303330, 0x36343030
```

alexanderwachter commented:
Thanks a lot @nandojve !

erwango commented on drivers/hwinfo/Kconfig:
```diff
@@ -21,55 +24,63 @@ config HWINFO_STM32
 	default y
 	depends on SOC_FAMILY_STM32
 	select USE_STM32_LL_UTILS
```
This could actually be removed since driver only require headers, not utils functions

alexanderwachter replied:
Thanks. Removed.

erwango reviewed (changes requested):
One minor related change requested, otherwise I'm fine with the change

erwango commented:
Before approving: I've tried to run sanitycheck, and there's a strange ERROR message reported (with no impact on final result):
```
$ sanitycheck -T tests/drivers/hwinfo/ -p nucleo_f401re
Renaming output directory to /local/mcu/zephyrproject/zephyr/sanity-out.2
INFO    - JOBS: 8
ERROR   - /local/mcu/zephyrproject/zephyr/tests/drivers/hwinfo/api/src/main.c: found invalid #endif, #ifdef in ztest_test_suite()
INFO    - Building initial testcase list...
INFO    - 1 test configurations selected, 0 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:    1/   1  100%  skipped:    0, failed:    0
INFO    - 1 of 1 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 4.94 seconds
INFO    - In total 1 test cases were executed on 1 out of total 229 platforms (0.44%)
```

Can you check the ERROR line ?

alexanderwachter commented:
@erwango pushed a new version that does not have the error above.
The fix is not to have a #ifdef in the test-suite but replace the tests with dummies depending on the configuration.

carlescufi commented:
@jhedberg this should be ready

carlescufi commented:
@jhedberg this should be ready

jhedberg commented:
> @jhedberg this should be ready

@carlescufi ok, I'll merge once CI completes.

alexanderwachter commented:
@jhedberg the test are timing out.  tests/drivers/hwinfo/api/drivers.device_id tests are all green.

jhedberg commented:
> @jhedberg the test are timing out. tests/drivers/hwinfo/api/drivers.device_id tests are all green.

Yes, I've had to restart Shippable a few times because of this. Fingers crossed that it wont time out again.
